src/main/java/bio/terra/service/filedata/VirtualFileSystemUtils.java

```diff
@@ -0,0 +1,157 @@
+                dirItem -> {
+                  final FireStoreFile file = fireStoreFiles.get(index.getAndIncrement());
+                  if (file == null) {
+                    throw new FileNotFoundException("File metadata was missing");
+                  }
+                  return dirItem
+                      .size(file.getSize())
+                      .checksumMd5(file.getChecksumMd5())
+                      .checksumCrc32c(file.getChecksumCrc32c());
+                }));
+      }
+    }
+  }
+
+  public static void updateDirectoryEntryChecksums(
+      FireStoreDirectoryEntry dirEntry, List<FireStoreDirectoryEntry> enumComputed) {
+    List<String> md5Collection = new ArrayList<>();
+    List<String> crc32cCollection = new ArrayList<>();
+    long totalSize = 0L;
+
+    for (FireStoreDirectoryEntry dirItem : enumComputed) {
+      totalSize = totalSize + dirItem.getSize();
+      if (!StringUtils.isEmpty(dirItem.getChecksumCrc32c())) {
+        crc32cCollection.add(dirItem.getChecksumCrc32c().toLowerCase());
+      }
+      if (!StringUtils.isEmpty(dirItem.getChecksumMd5())) {
+        md5Collection.add(dirItem.getChecksumMd5().toLowerCase());
+      }
+    }
+
+    // Compute checksums
+    // The spec is not 100% clear on the algorithm. I made specific choices on
+    // how to implement it:
+    // - set hex strings to lowercase before processing so we get consistent sort
+    //   order and digest results.
+    // - do not make leading zeros converting crc32 long to hex and it is returned
+    //   in lowercase. (Matches the semantics of toHexString).
+    Collections.sort(md5Collection);
+    String md5Concat = StringUtils.join(md5Collection, StringUtils.EMPTY);
+    String md5Checksum = computeMd5(md5Concat);
+
+    Collections.sort(crc32cCollection);
+    String crc32cConcat = StringUtils.join(crc32cCollection, StringUtils.EMPTY);
+    String crc32cChecksum = computeCrc32c(crc32cConcat);
+
+    // Update the directory in place
+    dirEntry.checksumCrc32c(crc32cChecksum).checksumMd5(md5Checksum).size(totalSize);
+  }
+
+  public static String getFullPath(String dirPath, String name) {
+    // Originally, this was a method in FireStoreDirectoryEntry, but the Firestore client complained
+    // about it,
+    // because it was not a set/get for an actual class member. Very picky, that!
+    // There are three cases here:
+    // - the path and name are empty: that is the root. Full path is "/"
+    // - the path is "/" and the name is not empty: dir in the root. Full path is "/name"
+    // - the path is "/name" and the name is not empty: Full path is path + "/" + name
+    String path = StringUtils.EMPTY;
+    if (StringUtils.isNotEmpty(dirPath) && !StringUtils.equals(dirPath, "/")) {
+      path = dirPath;
+    }
+    return path + '/' + name;
+  }
+
+  public static String computeMd5(String input) {
+    return StringUtils.lowerCase(DigestUtils.md5Hex(input));
+  }
+
+  public static String computeCrc32c(String input) {
```

Comment:
Is it possible this might be null?  Wouldn't that cause a problem in the azure case where there isn't a value for this?  What is input supposed to represent?  Might be good to have some javadoc for this method and the MD5 one